nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -217,7 +217,10 @@ WaypointFollower::followWaypoints()
     feedback->current_waypoint = goal_index;
     action_server_->publish_feedback(feedback);
 
-    if (current_goal_status_.status == ActionStatus::FAILED) {
+    if (
+      current_goal_status_.status == ActionStatus::FAILED ||
+      current_goal_status_.status == ActionStatus::UNKNOWN)
```

Comment:
Is this really a failed case if unknown from this timeout? Seems like we should instead temporarily increase the timeout back to 15 minutes (@pepisg ) while working through a potential patch to rclcpp to better keep actively processing goals around.

Reply:
Exposed that as a parameter in #3787

Reply:
This PR is complementary with PR3787, that improves the situation and fixes some bugs, that 3787 does not. So both PR-s ought to be be merged (moreover, new parameters in RewrittenYaml I plan to use to comb-up Costmap Filters system tests).

Reply:
@AlexeyMerzlyakov I meant only the addition of `|| current_goal_status_.status == ActionStatus::UNKNOWN`. I thought this line was to mask the unknown return type problem? If we increase the timeout, wouldn't that make it so we didn't see `UNKNOWN`? This logic treats `UNKNOWN` as a failure.

If we keep this here should https://github.com/ros-planning/navigation2/pull/3785/files#diff-872945d6a63001626ab7379639b5b3d5f11982ea3bdf08da01e7f111a9c2dfecR279 be changed for incrementing new goals? You're treating UNKNOWN as a terminal condition in your change, so this doesn't make as much sense anymore, correct?

Reply:
This status is being treated as terminal one for WPF server in order to avoid infinite cycles whether for some reasons the goal got into "UNKNOWN" state. This is intended to fix/mask not only the case when the goal is getting after it was destroyed, but rather for any other potential problems. Briefly looking to RCLCPP and RCL code shows that UNKNOWN state might appear in case if goal was invalidated, destroyed, does not exists at the moment of if the transition was invalid or any other error occurred. So, receiving UNKNOWN goal for action server seems to be treated as problem state, if I got the RCLCPP / RCL intention correctly.

When I've wrote this change, I've also focused on `BtActionNode` server, receiving the result from downstream nodes. For this case, "UNKNOWN" status is [being treated](https://github.com/ros-planning/navigation2/blob/be4d40711bbaf811aa3a0d5de346216ce67d0083/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L289) as error situation, and error is being thrown, interrupting normal operation.
From this point of view, looking on the UNKNOWN goal status as on error seems to be logical.

> If we keep this here should https://github.com/ros-planning/navigation2/pull/3785/files#diff-872945d6a63001626ab7379639b5b3d5f11982ea3bdf08da01e7f111a9c2dfecR279 be changed for incrementing new goals?

Yes, this should be removed for UNKNOWN stage, if we will treat UNKNOWNS as errors. Thanks for the notice!
If we will agree on main idea, I will remove it in next update.

Reply:
Yup that works for me!

Reply:
Done in https://github.com/ros-planning/navigation2/pull/3785/commits/85b6b71f6793f6740db7cd0d4d930bd19de3ca38